ja/console-and-shells.rst

```diff
@@ -1,18 +1,1094 @@
+.. php:method:err($message, $newlines)
+
+``Shell`` クラスは、コンテンツを出力するためのいくつかのメソッドを提供します。 ::
+
+    // 標準出力 (stdout) に出力
+    $this->out('Normal message');
+
+    // 標準エラー (stderr) に出力
+    $this->err('Error message');
+
+    // 標準エラー (stderr) に出力し、停止例外を発生
+    $this->abort('Fatal error');
+
+    // CakePHP 3.2 より前。標準エラー (stderr) に出力し exit() 
+    $this->error('Fatal error');
+
+また、出力レベルに関する2つの便利なメソッドを提供します。 ::
+
+    // 詳細出力が有効の時のみ (-v)
+    $this->verbose('Verbose message');
+
+    // すべてのレベルで表示
+    $this->quiet('Quiet message');
+
+シェルはまた、画面のクリア、空白行の作成、または横棒線を描くためのメソッドを含みます。 ::
+
+    // ２行の改行を出力
+    $this->out($this->nl(2));
+
+    // ユーザーの画面をクリア
+    $this->clear();
+
+    // 横棒線を描画
+    $this->hr();
+
+最後に、``_io->overwrite()`` を使用して、画面上のテキストの現在の行を更新することができます。 ::
+
+    $this->out('Counting down');
+    $this->out('10', 0);
+    for ($i = 9; $i > 0; $i--) {
+        sleep(1);
+        $this->_io->overwrite($i, 0, 2);
+    }
+
+新しい行が出力されたら、テキストを上書きすることができないことに注意してください。
+
+.. _shell-output-level:
+
+コンソール出力のレベル
+----------------------
+
+シェルは、往々にして異なる詳細レベルの出力を必要とします。cron ジョブとして実行している場合は
+ほとんどの出力は不要です。また、シェルが出力する メッセージを見たくないというケースもあるでしょう。
+このような場合、出力を適切に制御するための出力レベルが使えます。シェルの利用者は、シェルを呼び出す際に
+正しいフラグをセットすることで、自分が見たい詳細レベルの設定 ができます。
+:php:meth:`Cake\\Console\\Shell::out()` では、デフォルトで３種類の出力をサポートしています。
+
+* ``QUIET`` - 必須のメッセージであり、静かな（必要最小限の）出力モードでも表示される。
+* ``NORMAL`` - 通常利用におけるデフォルトのレベル。
+* ``VERBOSE`` - 毎日利用には冗長すぎるメッセージを表示、しかしデバッグ時には有用。
+
+以下のように出力を指定できます。 ::
+
+    // すべてのレベルで表示されます。
+    $this->out('Quiet message', 1, Shell::QUIET);
+    $this->quiet('Quiet message');
+
+    // QUIET 出力時には表示されません。
+    $this->out('normal message', 1, Shell::NORMAL);
+    $this->out('loud message', 1, Shell::VERBOSE);
+    $this->verbose('Verbose output');
+
+    // VERBOSE 出力時のみ表示されます。
+    $this->out('extra message', 1, Shell::VERBOSE);
+    $this->verbose('Verbose output');
+
+シェルの実行時に ``--quiet`` や ``--verbose`` を使うことで出力を制御できます。
+これらのオプションはデフォルトで組み込まれていて、いつでも CakePHP のシェル内部の
+出力レベルを制御できるように考慮されています。
+
+また、 ``--quiet`` と ``--verbose`` オプションは、ログデータの標準出力/標準エラーへの
+出力方法を制御します。通常の情報とそれ以上のレベルのログメッセージは標準出力/標準エラーに出力されます。
+``--verbose`` を使用する場合は、デバッグログは標準出力に出力されます。
+``--quiet`` を使用する場合は、警告とそれ以上のレベルのログメッセージのみ標準エラーに出力されます。
+
+出力のスタイル
+--------------
+
+ちょうど HTML のようなタグを埋め込むことで、出力のスタイルを変更することができます。
+ConsoleOutput はこれらのタグを正しい ansi コードシーケンスに変換したり、ansi コードを
+サポートしないコンソールではタグを除去します。
+スタイルはいくつかビルトインされたものがありますが、自分で作成することも 可能です。
+ビルトインされたものは以下の通りです。
+
+* ``error`` エラーメッセージ。赤色のテキスト。
+* ``warning`` 警告メッセージ。黄色のテキスト。
+* ``info`` 情報メッセージ。水色のテキスト。
+* ``comment`` 追加情報。青色のテキスト。
+* ``question`` 質問事項。シェルが自動的に追加する。
+
+``$this->stdout->styles()`` を使ってさらに多くのスタイルを追加できます。
+新しいスタイルを追加するには以下のようにします。 ::
+
+    $this->_io->styles('flashy', ['text' => 'magenta', 'blink' => true]);
+
+これで ``<flashy>`` というタグが有効になり、ansi カラーが有効な端末であれば、
+``$this->out('<flashy>うわ！</flashy> 何か変になった');`` の場合の表示は
+色がマゼンタでブリンクになります。
+スタイルを定義する際は `text` と `background` 属性として以下の色が指定できます。
```

Comment:
In the English version, `text` and `background` are enclosed with double backquotes.